dashboard/src/shared/schema.ts

```diff
@@ -1,16 +1,74 @@
+// WARN: yaml doesn't have updated definitions for TypeScript
+// In particular, it doesn't contain definitions for `get` and `set`
+// that are used in this package
+import * as jsonSchema from "json-schema";
+import * as YAML from "yaml";
 import { IBasicFormParam } from "./types";
 
-export function retrieveBasicFormParams(schema: any) {
-  // TBD
-  return {
-    username: {
-      path: "wordpressUsername",
-      value: "user",
-    } as IBasicFormParam,
-  };
+// Avoid to explicitly add "null" when an element is not defined
+// tslint:disable-next-line
+const { nullOptions } = require("yaml/types");
+nullOptions.nullStr = "";
+
+// retrieveBasicFormParams iterates over a JSON Schema properties looking for `form` keys
+// It uses the raw yaml to setup default values.
+// It returns a key:value map for easier handling.
+export function retrieveBasicFormParams(
+  defaultValues: string,
+  schema: jsonSchema.JSONSchema4,
+): { [key: string]: IBasicFormParam } {
+  if (schema && schema.properties) {
+    return lookForFormParams(defaultValues, schema.properties);
+  }
+  return {};
 }
 
+// setValue modifies the current values (text) based on a path
 export function setValue(values: string, path: string, newValue: any) {
-  // TBD
-  return values;
+  const doc = YAML.parseDocument(values);
+  const splittedPath = path.split(".");
+  (doc as any).setIn(splittedPath, newValue);
+  return doc.toString();
+}
+
+// getValue returns the current value of an object based on YAML text and its path
+export function getValue(values: string, path: string, defaultValue?: any) {
+  const doc = YAML.parseDocument(values);
+  const splittedPath = path.split(".");
+  return (doc as any).getIn(splittedPath) || defaultValue;
+}
```

Comment:
Great - `getValue` and `setValue` are much simpler now. I'm still not sure why we pass string values into these, rather than passing the actual document. I expected that the form state would be the document, which gets updated with data by the user... converting that to a string and back on each change seems superfluous, but maybe there is context I'm missing?

Reply:
My reasoning is that the the `values.yaml` is received as a string and it's always treated as a normal string in the `Form` component if it doesn't have support for the basic form.

My idea is to have this abstraction layer so the `Form` always work with the string `values.yaml` regardless of the implementation of this library. We are using yaml `Document`s but we could be using regexps or `lodash` to modify those values. That way, the `Form` component doesn't depend on the implementation details here.